### PR TITLE
zephyrCommon: Implement pulseIn

### PR DESCRIPTION
**Pull request by Ayush1325**
- A basic sync implementation

### CODE REVIEW

File: cores/arduino/zephyrCommon.cpp [hunk 1]
```diff
@@ -328,3 +328,28 @@ void detachInterrupt(pin_size_t pinNumber)
...
+
+
+	while(gpio_pin_get_dt(spec) == state);
+	while(gpio_pin_get_dt(spec) != state);
+
+	start = k_uptime_ticks();
+	while(gpio_pin_get_dt(spec) == state);
+	end = k_uptime_ticks();
+
+	delta = k_ticks_to_us_floor64(end) - k_ticks_to_us_floor64(start);
```
Comment (mathieuchopstm) [inline cores/arduino/zephyrCommon.cpp]: Calculating delta in ticks would reduce rounding errors (`delta = k_ticks_to_us_floor64(end - start);`)

File: cores/arduino/zephyrCommon.cpp [hunk 2]
```diff
@@ -328,3 +328,28 @@ void detachInterrupt(pin_size_t pinNumber)
...
+
+	if (!gpio_is_ready_dt(spec)) {
+		return 0;
+	}
+
+	if (!gpio_pin_is_input_dt(spec)) {
+		return 0;
+	}
+
+
```
Comment (DhruvaG2000) [inline cores/arduino/zephyrCommon.cpp:370]: remove extra blank line

File: cores/arduino/zephyrCommon.cpp [hunk 3]
```diff
@@ -328,3 +328,28 @@ void detachInterrupt(pin_size_t pinNumber)
...
+	while(gpio_pin_get_dt(spec) == state);
+	while(gpio_pin_get_dt(spec) != state);
```
Comment (DhruvaG2000) [inline cores/arduino/zephyrCommon.cpp]: There's no timeout logic?

**Review by DhruvaG2000 — CHANGES_REQUESTED** [review]

File: cores/arduino/zephyrCommon.cpp (hunk 3 above)
Comment (Ayush1325) [reply cores/arduino/zephyrCommon.cpp]: Right, I forgot to add that.

Comment (Ayush1325) [reply cores/arduino/zephyrCommon.cpp]: I have added a simple timer-based implementation.

File: cores/arduino/zephyrCommon.cpp [hunk 4]
```diff
@@ -349,3 +349,50 @@ long random(long max) {
...
+  end = k_uptime_ticks();
+
+  delta = k_ticks_to_us_floor64(end - start);
+
+cleanup:
+  k_timer_stop(&timer);
+  return (unsigned long)delta;
+}
+
+#endif
```
Comment (DhruvaG2000) [inline cores/arduino/zephyrCommon.cpp]: I know we've missed adding this in the past, but please add a comment to identify the endif:
```
// CONFIG_GPIO_GET_DIRECTION
```

File: cores/arduino/zephyrCommon.cpp [hunk 5]
```diff
@@ -349,3 +349,50 @@ long random(long max) {
...
+  while(gpio_pin_get_dt(spec) != state) {
+    if (k_timer_status_get(&timer) > 0) {
+      goto cleanup;
+    }
+  }
+
+  start = k_uptime_ticks();
+  while(gpio_pin_get_dt(spec) == state) {
+    if (k_timer_status_get(&timer) > 0) {
+      goto cleanup;
```
Comment (DhruvaG2000) [inline cores/arduino/zephyrCommon.cpp]: Here and everywhere else above, I am not a big fan of using goto's inside while loops please.

Comment (DhruvaG2000) [reply cores/arduino/zephyrCommon.cpp]: I'd rather you break the loop and then add a conditional goto cleanup outside it.

**Review by DhruvaG2000 — CHANGES_REQUESTED** [review]
Mostly looks good, with the minor suggestions we should be ok.

**Comment by Ayush1325** [comment]
@DhruvaG2000 Any idea regarding the failing CI? I don't think it is related to this PR.

**Comment by DhruvaG2000** [comment]
> @DhruvaG2000 Any idea regarding the failing CI?

Yep, locally reproduced just now.
``/__w/gsoc-2022-arduino-core/gsoc-2022-arduino-core/Arduino-Zephyr-API/samples/i2cdemo/prj.conf:2: warning: attempt to assign the value 'y' to the undefined symbol CPLUSPLUS``

Need to get rid of this symbol from everywhere looks like.

**Comment by DhruvaG2000** [comment]
Not related to this PR, so nothing to be alarmed.

**Comment by DhruvaG2000** [comment]
This PR should probably help:
https://github.com/zephyrproject-rtos/gsoc-2022-arduino-core/pull/104

**Review by DhruvaG2000 — APPROVED** [review]

**Pull request merged**